src/main/scala/io/iohk/ethereum/jsonrpc/server/http/RateLimit.scala

```diff
@@ -1,25 +1,67 @@
-import io.iohk.ethereum.jsonrpc.server.http.JsonRpcHttpServer.JsonRpcHttpServerConfig
+import akka.NotUsed
+import akka.http.scaladsl.model.{RemoteAddress, StatusCodes}
+import akka.http.scaladsl.server.{Directive0, Route}
+import io.iohk.ethereum.jsonrpc.server.http.JsonRpcHttpServer.RateLimitConfig
+import akka.http.scaladsl.server.Directives._
+import com.google.common.base.Ticker
+import com.google.common.cache.CacheBuilder
+import io.iohk.ethereum.jsonrpc.JsonRpcError
+import de.heikoseeberger.akkahttpjson4s.Json4sSupport
+import io.iohk.ethereum.jsonrpc.serialization.JsonSerializers
+import org.json4s.{DefaultFormats, Formats, Serialization, native}
 
-trait RateLimit {
+class RateLimit(config: RateLimitConfig) extends Directive0 with Json4sSupport {
 
-  val config: JsonRpcHttpServerConfig
+  private implicit val serialization: Serialization = native.Serialization
+  private implicit val formats: Formats = DefaultFormats + JsonSerializers.RpcErrorJsonSerializer
 
-  val latestRequestTimestamps = new LruMap[RemoteAddress, Long](config.rateLimit.latestTimestampCacheSize)
+  protected def getCurrentTimeNanos: Long = System.nanoTime()
 
-  val clock: Clock = Clock.systemUTC()
-
-  def isBelowRateLimit(clientAddress: RemoteAddress): Boolean = {
-    val timeMillis = clock.instant().toEpochMilli
-    val latestRequestTimestamp = latestRequestTimestamps.getOrElse(clientAddress, 0L)
+  private[this] lazy val lru = {
+    val nanoDuration = config.minRequestInterval.toNanos
+    val javaDuration = Duration.ofNanos(nanoDuration)
+    val ticker: Ticker = new Ticker {
+      override def read(): Long = getCurrentTimeNanos
+    }
+    CacheBuilder
+      .newBuilder()
+      .weakKeys()
+      .expireAfterAccess(javaDuration)
+      .ticker(ticker)
+      .build[RemoteAddress, NotUsed]()
+  }
 
-    val response = latestRequestTimestamp + config.rateLimit.minRequestInterval.toMillis < timeMillis
-    if (response) latestRequestTimestamps.put(clientAddress, timeMillis)
-    response
+  // Such algebras prevent if-elseif-else boilerplate in the JsonRPCServer code
+  // It is also guaranteed that:
+  //   1) no IP address is extracted unless config.enabled is true
+  //   2) no LRU is created unless config.enabled is true
+  //   3) cache is accessed only once (using get)
+  override def tapply(f: Unit => Route): Route = {
+    if (config.enabled) {
+      val minInterval = config.minRequestInterval.toSeconds
+        extractClientIP { ip =>
+          var exists = true
+          // We can avoid using var
+          // But in this case we access our LRU twice.
+          lru.get(
+            ip,
+            () => {
+              exists = false
+              NotUsed
+            }
+          )
+          if (exists) {
+            val err = JsonRpcError.RateLimitError(minInterval)
+            complete((StatusCodes.TooManyRequests, err))
+          } else {
+            f.apply(())
+          }
+        }
+      } else f.apply(())
   }
+
```

Comment:
Maybe we can encapsulate the validation behavior, for example:

private def isBelowRateLimit(ip: RemoteAddress): Boolean = {
    var exists = true
    lru.get(
      ip,
      () => {
        exists = false
        NotUsed
      }
    )
    exists
  }

And then call this method:
extractClientIP { ip =>
 if (isBelowRateLimit(ip)) {

Reply:
It's done